In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import pickle as p

# to display all columns and rows:
pd.set_option('display.max_columns', None); 
pd.set_option('display.max_rows', None);

# virgulden sonra gösterilecek olan sayı sayısı
pd.set_option('display.float_format', lambda x: '%.2f' % x)
import matplotlib.pyplot as plt

In [3]:
pwd

'C:\\Users\\merta\\notebooks\\odevler'

In [4]:
# 2009-2010 years data
df_010_2011 = pd.read_excel("C:\\Users\\merta\\data\\online_retail_II.xlsx", sheet_name = "Year 2010-2011")

In [5]:
df = df_2010_2011.copy() # do all things on df and backup first data

In [6]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom

In [7]:
df.tail()

Invoice StockCode                      Description  Quantity  \
541905  581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906  581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907  581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908  581587     22138    BAKING SET 9 PIECE RETROSPOT          3   
541909  581587      POST                          POSTAGE         1   

               InvoiceDate  Price  Customer ID Country  
541905 2011-12-09 12:50:00   2.10     12680.00  France  
541906 2011-12-09 12:50:00   4.15     12680.00  France  
541907 2011-12-09 12:50:00   4.15     12680.00  France  
541908 2011-12-09 12:50:00   4.95     12680.00  France  
541909 2011-12-09 12:50:00  18.00     12680.00  France

In [8]:
df.shape

(541910, 8)

In [9]:
df.shape[0] #satir sayisi

541910

In [10]:
df.shape[1] #sutun sayisi

8

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[ns]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [12]:
df.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object')

In [13]:
df.index

RangeIndex(start=0, stop=541910, step=1)

In [14]:
df.isnull().values.any() #Hiç Eksik deger var mı?

True

In [15]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

In [17]:
df.describe().T #descriptive statistics

count     mean     std       min      25%      50%      75%  \
Quantity    541910.00     9.55  218.08 -80995.00     1.00     3.00    10.00   
Price       541910.00     4.61   96.76 -11062.06     1.25     2.08     4.13   
Customer ID 406830.00 15287.68 1713.60  12346.00 13953.00 15152.00 16791.00   

                 max  
Quantity    80995.00  
Price       38970.00  
Customer ID 18287.00

In [18]:
#unique product count
df["Description"].nunique()

4223

In [19]:
df["Description"].value_counts().head()

WHITE HANGING HEART T-LIGHT HOLDER    2369
REGENCY CAKESTAND 3 TIER              2200
JUMBO BAG RED RETROSPOT               2159
PARTY BUNTING                         1727
LUNCH BAG RED RETROSPOT               1638
Name: Description, dtype: int64

In [20]:
df.groupby("Description").agg({"Quantity":"sum"}).sort_values("Quantity", ascending = False).head()

Quantity
Description                                
WORLD WAR 2 GLIDERS ASSTD DESIGNS     53847
JUMBO BAG RED RETROSPOT               47363
ASSORTED COLOUR BIRD ORNAMENT         36381
POPCORN HOLDER                        36334
PACK OF 72 RETROSPOT CAKE CASES       36039

In [21]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom

In [23]:
df["TotalPrice"] = df["Price"] * df["Quantity"] #TotalPrice calculation

In [24]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [26]:
df[df['Invoice'].str.contains('C', na = False)].head() # Cannceled invoices

Invoice StockCode                       Description  Quantity  \
141  C536379         D                          Discount        -1   
154  C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235  C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236  C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237  C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

            InvoiceDate  Price  Customer ID         Country  TotalPrice  
141 2010-12-01 09:41:00  27.50     14527.00  United Kingdom      -27.50  
154 2010-12-01 09:49:00   4.65     15311.00  United Kingdom       -4.65  
235 2010-12-01 10:24:00   1.65     17548.00  United Kingdom      -19.80  
236 2010-12-01 10:24:00   0.29     17548.00  United Kingdom       -6.96  
237 2010-12-01 10:24:00   0.29     17548.00  United Kingdom       -6.96

In [27]:
df = df[~df["Invoice"].str.contains("C", na = False)] #data uncancelled

In [28]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [30]:
df.groupby(df["Invoice"]).agg({"TotalPrice":'sum'}).head() #totalPrices of Invoices

TotalPrice
Invoice            
536365       139.12
536366        22.20
536367       278.73
536368        70.05
536369        17.85

df["Country"].value_counts() #Country base invoice counts

In [35]:
df.groupby(df["Country"]).agg({"TotalPrice":'sum'}).head()#country based sum of TotalPrices

TotalPrice
Country              
Australia   138521.31
Austria      10198.68
Bahrain        754.14
Belgium      41196.34
Brazil        1143.60

In [36]:
df.groupby(df["Country"]).agg({"TotalPrice":'sum'}).sort_values("TotalPrice", ascending = False).head() # sorted

TotalPrice
Country                   
United Kingdom  9003097.96
Netherlands      285446.34
EIRE             283453.96
Germany          228867.14
France           209733.11

In [37]:
df1 = df_2010_2011.copy()

In [38]:
df1.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom

In [39]:
df1[df1["Invoice"].str.contains("C", na = False)].head() #canceled invoices

Invoice StockCode                       Description  Quantity  \
141  C536379         D                          Discount        -1   
154  C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235  C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236  C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237  C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

            InvoiceDate  Price  Customer ID         Country  
141 2010-12-01 09:41:00  27.50     14527.00  United Kingdom  
154 2010-12-01 09:49:00   4.65     15311.00  United Kingdom  
235 2010-12-01 10:24:00   1.65     17548.00  United Kingdom  
236 2010-12-01 10:24:00   0.29     17548.00  United Kingdom  
237 2010-12-01 10:24:00   0.29     17548.00  United Kingdom

In [41]:
df2 = df1[df1["Invoice"].str.contains("C", na = False)]

In [42]:
df2["Description"].value_counts().head() # Count of different product desc for cancelled invoices

Manual                      244
REGENCY CAKESTAND 3 TIER    181
POSTAGE                     126
JAM MAKING SET WITH JARS     87
Discount                     77
Name: Description, dtype: int64

In [43]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [44]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    134697
Country             0
TotalPrice          0
dtype: int64

In [45]:
df.dropna(inplace = True) # drop of nan values

In [46]:
df.shape

(397925, 9)

In [48]:
df.describe([0.01,0.05,0.10,0.25,0.50,0.75,0.90,0.95, 0.99]).T #descriptive statistics for quantiles we provided

count     mean     std      min       1%       5%      10%  \
Quantity    397925.00    13.02  180.42     1.00     1.00     1.00     1.00   
Price       397925.00     3.12   22.10     0.00     0.21     0.42     0.55   
Customer ID 397925.00 15294.31 1713.17 12346.00 12415.00 12627.00 12883.00   
TotalPrice  397925.00    22.39  309.06     0.00     0.55     1.25     1.95   

                 25%      50%      75%      90%      95%      99%       max  
Quantity        2.00     6.00    12.00    24.00    36.00   120.00  80995.00  
Price           1.25     1.95     3.75     6.35     8.50    14.95   8142.75  
Customer ID 13969.00 15159.00 16795.00 17725.00 17912.00 18211.00  18287.00  
TotalPrice      4.68    11.80    19.80    35.40    67.50   202.50 168469.60

In [50]:
for feature in ["Quantity","Price","TotalPrice"]: # outlier observation for quantiles

    Q1 = df[feature].quantile(0.01)
    Q3 = df[feature].quantile(0.99)
    IQR = Q3-Q1
    upper = Q3 + 1.5*IQR
    lower = Q1 - 1.5*IQR

    if df[(df[feature] > upper) | (df[feature] < lower)].any(axis=None):
        print(feature,"yes")
        print(df[(df[feature] > upper) | (df[feature] < lower)].shape[0])
    else:
        print(feature, "no")

Quantity yes
963
Price yes
661
TotalPrice yes
903


aykiri degerlere birsey yapmiyoruz henuz

# Recency

In [51]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 397925 entries, 0 to 541909
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      397925 non-null  object        
 1   StockCode    397925 non-null  object        
 2   Description  397925 non-null  object        
 3   Quantity     397925 non-null  int64         
 4   InvoiceDate  397925 non-null  datetime64[ns]
 5   Price        397925 non-null  float64       
 6   Customer ID  397925 non-null  float64       
 7   Country      397925 non-null  object        
 8   TotalPrice   397925 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(4)
memory usage: 30.4+ MB


In [53]:
df["InvoiceDate"].min()

Timestamp('2010-12-01 08:26:00')

In [45]:
df["InvoiceDate"].max() #nearest invoice date

Timestamp('2011-12-09 12:50:00')

In [57]:
import datetime as dt

In [58]:
today_date = dt.datetime(2011, 12, 9) # assumed today's date is the nearest invoice date in data

In [59]:
today_date

datetime.datetime(2011, 12, 9, 0, 0)

In [49]:
df.groupby("Customer ID").agg({"InvoiceDate":"max"}).head() # customers' nearest visiting dates

InvoiceDate
Customer ID                    
12346.00    2011-01-18 10:01:00
12347.00    2011-12-07 15:52:00
12348.00    2011-09-25 13:13:00
12349.00    2011-11-21 09:51:00
12350.00    2011-02-02 16:01:00

In [60]:
df["Customer ID"] = df["Customer ID"].astype(int) #type cast for customer ID

In [61]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55        17850  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75        17850  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34

In [62]:
df.groupby(df["Customer ID"]).agg({"InvoiceDate":'max'}).head()

InvoiceDate
Customer ID                    
12346       2011-01-18 10:01:00
12347       2011-12-07 15:52:00
12348       2011-09-25 13:13:00
12349       2011-11-21 09:51:00
12350       2011-02-02 16:01:00

In [65]:
today_date - df.groupby(df["Customer ID"]).agg({"InvoiceDate":'max'}) 
# broadcasting feature has been used for subtracting nearest invoice date from today's date

InvoiceDate
Customer ID                  
12346       324 days 13:59:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00
12352        35 days 09:23:00
12353       203 days 06:13:00
12354       231 days 10:49:00
12355       213 days 10:11:00
12356        21 days 15:20:00
12357        32 days 07:53:00
12358         0 days 13:34:00
12359        56 days 11:13:00
12360        51 days 08:38:00
12361       286 days 10:09:00
12362         2 days 08:20:00
12363       108 days 13:42:00
12364         6 days 13:38:00
12365       290 days 09:56:00
12367         3 days 07:12:00
12370        50 days 09:09:00
12371        43 days 13:44:00
12372        70 days 11:48:00
12373       310 days 10:50:00
12374        24 days 08:23:00
12375         9 days 13:24:00
12377       314 days 08:15:00
12378       128 days 13:26:00
12379        80 days 13:51:00
12380        20 days 12:33:00
12381         3 days 11:16:00
12383       183 days 15:58:00
12384        27 days 15:21:00
12386       336 days 11:23:00
12388        14 days 11:30:00
12390        78 days 14:27:00
12391        20 days 11:53:00
12393        71 days 08:19:00
12394        62 days 15:52:00
12395        18 days 08:39:00
12397        34 days 14:08:00
12398        44 days 13:33:00
12399       118 days 10:45:00
12401       302 days 10:10:00
12402       322 days 10:11:00
12403        48 days 13:09:00
12405       147 days 15:08:00
12406        21 days 04:38:00
12407        48 days 13:49:00
12408        31 days 11:33:00
12409        77 days 13:22:00
12410       307 days 13:21:00
12412        73 days 13:44:00
12413        65 days 15:00:00
12414       216 days 09:09:00
12415        23 days 09:38:00
12417         2 days 09:08:00
12418       111 days 12:52:00
12420        62 days 15:05:00
12421        14 days 14:53:00
12422        94 days 14:12:00
12423       -1 days +13:50:00
12424       161 days 11:54:00
12425        77 days 06:29:00
12426       193 days 11:34:00
12427        20 days 15:04:00
12428        24 days 15:38:00
12429         8 days 06:38:00
12430        42 days 08:51:00
12431        34 days 12:05:00
12432        41 days 12:17:00
12433       -1 days +13:58:00
12434        83 days 11:22:00
12435        78 days 14:54:00
12436        98 days 11:51:00
12437         0 days 12:25:00
12438        13 days 10:32:00
12441       365 days 13:15:00
12442         2 days 09:38:00
12444        20 days 10:37:00
12445        21 days 05:17:00
12446        56 days 11:10:00
12447       242 days 10:35:00
12448        43 days 10:14:00
12449        21 days 10:14:00
12450       155 days 09:19:00
12451         9 days 15:20:00
12452        15 days 13:32:00
12453       133 days 09:52:00
12454        55 days 06:47:00
12455        72 days 08:29:00
12456        43 days 09:19:00
12457        65 days 10:47:00
12458        70 days 08:35:00
12461        93 days 07:03:00
12462        17 days 14:49:00
12463        52 days 10:16:00
12464         9 days 10:27:00
12465        16 days 07:26:00
12468       186 days 12:24:00
12471         1 days 08:17:00
12472        31 days 09:59:00
12473        28 days 10:14:00
12474        16 days 09:21:00
12475        52 days 13:50:00
12476        14 days 11:20:00
12477        21 days 14:09:00
12478         2 days 14:04:00
12479        10 days 10:34:00
12480        27 days 09:08:00
12481        21 days 15:31:00
12483        16 days 06:58:00
12484        31 days 12:23:00
12488         8 days 15:10:00
12489       335 days 10:47:00
12490         4 days 07:53:00
12491        37 days 11:42:00
12492        62 days 08:45:00
12493       164 days 13:43:00
12494        14 days 14:26:00
12497        80 days 11:31:00
12498        20 days 11:07:00
12500        22 days 07:40:00
12501       335 days 11:32:00
12502        94 days 12:08:00
12504        17 days 07:58:00
12506       231 days 15:06:00
12507       133 days 06:51:00
12508        25 days 08:23:00
12509       283 days 09:15:00
12510       141 days 13:45:00
12512        65 days 09:05:0

In [66]:
today_date

datetime.datetime(2011, 12, 9, 0, 0)

In [71]:
temp_df = (today_date - df.groupby("Customer ID").agg({"InvoiceDate":"max"}))

In [72]:
temp_df.head()

InvoiceDate
Customer ID                  
12346       324 days 13:59:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00

In [73]:
temp_df.rename(columns = {"InvoiceDate":"Recency"}, inplace = True) # Recency for all customers

In [74]:
temp_df.head()

Recency
Customer ID                  
12346       324 days 13:59:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00

In [75]:
temp_df.iloc[0,0].days #only day value is needed

324

In [76]:
recency_df = temp_df["Recency"].apply(lambda x: x.days) #day value calculation for all customers

In [77]:
recency_df.head()

Customer ID
12346    324
12347      1
12348     74
12349     17
12350    309
Name: Recency, dtype: int64

# Frequency


In [78]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55        17850  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75        17850  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34

In [79]:
df.groupby(["Customer ID","Invoice"]).agg({"Invoice":"nunique"}).head(50)

Invoice
Customer ID Invoice         
12346       541431         1
12347       537626         1
            542237         1
            549222         1
            556201         1
            562032         1
            573511         1
            581180         1
12348       539318         1
            541998         1
            548955         1
            568172         1
12349       577609         1
12350       543037         1
12352       544156         1
            545323         1
            545332         1
            546869         1
            547390         1
            567505         1
            568699         1
            574275         1
12353       553900         1
12354       550911         1
12355       552449         1
12356       541430         1
            549435         1
            576895         1
12357       574740         1
12358       559707         1
            581232         1
12359       540946         1
            543370         1
            555477         1
            571034         1
12360       554132         1
            563749         1
            571705         1
12361       545017         1
12362       544203         1
            551346         1
            559295         1
            563037         1
            568651         1
            570667         1
            572887         1
            573173         1
            574329         1
            580979         1
12363       549687         1

In [80]:
freq_df = df.groupby("Customer ID").agg({"InvoiceDate":"nunique"})# unique visit date counts of customers (frequencies)

In [81]:
freq_df.head()

InvoiceDate
Customer ID             
12346                  1
12347                  7
12348                  4
12349                  1
12350                  1

In [82]:
freq_df.rename(columns={"InvoiceDate": "Frequency"}, inplace=True)

# MONETARY

In [83]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55        17850  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75        17850  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34

In [85]:
monetary_df = df.groupby("Customer ID").agg({"TotalPrice":"sum"}) # monetary values per customers

In [88]:
monetary_df.head()

Monetary
Customer ID          
12346        77183.60
12347         4310.00
12348         1797.24
12349         1757.55
12350          334.40

In [89]:
monetary_df.rename(columns={"TotalPrice":"Monetary"}, inplace=True)

In [90]:
monetary_df.head()

Monetary
Customer ID          
12346        77183.60
12347         4310.00
12348         1797.24
12349         1757.55
12350          334.40

In [91]:
print(recency_df.shape,freq_df.shape,monetary_df.shape)

(4339,) (4339, 1) (4339, 1)


In [92]:
rfm = pd.concat([recency_df, freq_df, monetary_df],  axis=1) #rfm dataframe

In [93]:
rfm.head()

Recency  Frequency  Monetary
Customer ID                              
12346            324          1  77183.60
12347              1          7   4310.00
12348             74          4   1797.24
12349             17          1   1757.55
12350            309          1    334.40

In [96]:
rfm["RecencyScore"] = pd.qcut(rfm["Recency"], 5, labels = [5, 4 , 3, 2, 1]) 
#sorted and calculated 5 quantile values for Recency  and labeled (descending, this is important!)

In [97]:
rfm.head()

Recency  Frequency  Monetary RecencyScore
Customer ID                                           
12346            324          1  77183.60            1
12347              1          7   4310.00            5
12348             74          4   1797.24            2
12349             17          1   1757.55            4
12350            309          1    334.40            1

In [102]:
#rfm["FrequencyScore"] = pd.qcut(rfm['Frequency'], 5, labels = [1, 2, 3, 4, 5]) 
#rank to take first value of a quantile in order to handle error

In [103]:
rfm["FrequencyScore"]= pd.qcut(rfm["Frequency"].rank(method="first"),5, labels=[1,2,3,4,5])
#sorted and calculated 5 quantile values for Frequency  and labeled (ascending, this is important!)

In [104]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore  
Customer ID                
12346                   5  
12347                   5  
12348                   4  
12349                   4  
12350                   2

In [108]:
rfm["MonetaryScore"] = pd.qcut(rfm['Monetary'], 5, labels = [1, 2, 3, 4, 5])
#sorted and calculated 5 quantile values for Monetary  and labeled (ascending, this is important!)

In [109]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore  
Customer ID                
12346                   5  
12347                   5  
12348                   4  
12349                   4  
12350                   2

In [110]:
(rfm['RecencyScore'].astype(str) + 
 rfm['FrequencyScore'].astype(str) + 
 rfm['MonetaryScore'].astype(str)).head()

Customer ID
12346    115
12347    555
12348    244
12349    414
12350    112
dtype: object

In [111]:
# rfm skorları kategorik değere dönüştürülüp df'e eklendi
rfm["RFM_SCORE"] = (rfm['RecencyScore'].astype(str) + 
                    rfm['FrequencyScore'].astype(str) + 
                    rfm['MonetaryScore'].astype(str)) # RFM score calculation

In [112]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE  
Customer ID                          
12346                   5       115  
12347                   5       555  
12348                   4       244  
12349                   4       414  
12350                   2       112

In [113]:
rfm[rfm["RFM_SCORE"]=="555"].head() #champions!!!

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12347              1          7   4310.00            5              5   
12362              2         10   5226.23            5              5   
12417              2          9   3649.10            5              5   
12433             -1          7  13375.87            5              5   
12437              0         18   4951.41            5              5   

            MonetaryScore RFM_SCORE  
Customer ID                          
12347                   5       555  
12362                   5       555  
12417                   5       555  
12433                   5       555  
12437                   5       555

In [86]:
rfm.describe().T

count    mean     std   min    25%    50%     75%       max
Recency   4339.00   91.04  100.01 -1.00  16.00  49.00  140.50    372.00
Frequency 4339.00    4.25    7.64  1.00   1.00   2.00    5.00    211.00
Monetary  4339.00 2053.80 8988.25  0.00 307.24 674.45 1661.64 280206.02

In [114]:
rfm[rfm["RFM_SCORE"] == "111"].head() # hibernating!!!

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12353            203          1     89.00            1              1   
12361            286          1    189.90            1              1   
12401            302          1     84.30            1              1   
12402            322          1    225.60            1              1   
12441            365          1    173.55            1              1   

            MonetaryScore RFM_SCORE  
Customer ID                          
12353                   1       111  
12361                   1       111  
12401                   1       111  
12402                   1       111  
12441                   1       111

In [118]:
# RFM Map via Regular Expressions
seg_map = {
    r'[1-2][1-2]': 'Hibernating',
    r'[1-2][3-4]': 'At Risk',
    r'[1-2]5': 'Can\'t Loose',
    r'3[1-2]': 'About to Sleep',
    r'33': 'Need Attention',
    r'[3-4][4-5]': 'Loyal Customers',
    r'41': 'Promising',
    r'51': 'New Customers',
    r'[4-5][2-3]': 'Potential Loyalists',
    r'5[4-5]': 'Champions'
}

In [119]:
rfm['Segment'] = rfm['RecencyScore'].astype(str) + rfm['FrequencyScore'].astype(str)
#we'll use only Recency and Frequency scores to calculate RFM Scores

In [120]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE Segment  
Customer ID                                  
12346                   5       115      11  
12347                   5       555      55  
12348                   4       244      24  
12349                   4       414      41  
12350                   2       112      11

In [121]:
rfm['Segment'] = rfm['Segment'].replace(seg_map, regex=True)# replate Segment values with Regex map

In [122]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE      Segment  
Customer ID                                       
12346                   5       115  Hibernating  
12347                   5       555    Champions  
12348                   4       244      At Risk  
12349                   4       414    Promising  
12350                   2       112  Hibernating

In [123]:
rfm[["Segment","Recency","Frequency", "Monetary"]].groupby("Segment").agg(["mean","median","count"])

Recency              Frequency              Monetary  \
                       mean median count      mean median count     mean   
Segment                                                                    
About to Sleep        51.37     51   353      1.16      1   353   472.55   
At Risk              152.04    137   594      2.87      3   594  1079.81   
Can't Loose          130.05    105    64      8.31      7    64  2791.01   
Champions              4.37      3   632     12.34      8   632  6866.78   
Hibernating          215.66    217  1069      1.10      1  1069   488.86   
Loyal Customers       31.69     28   820      6.44      5   820  2862.89   
Need Attention        50.27     50   184      2.32      2   184   894.49   
New Customers          5.43      6    42      1.00      1    42   388.21   
Potential Loyalists   15.37     16   486      2.01      2   486  1041.34   
Promising             21.42     21    95      1.00      1    95   290.91   

                                   
                     median count  
Segment                            
About to Sleep       331.95   353  
At Risk              675.16   594  
Can't Loose         2236.74    64  
Champions           2613.85   632  
Hibernating          293.00  1069  
Loyal Customers     1737.83   820  
Need Attention       637.78   184  
New Customers        274.50    42  
Potential Loyalists  524.00   486  
Promising            219.00    95

In [124]:
rfm.shape[0] #Total customer count

4339

In [127]:
rfm[rfm["Segment"] == "Need Attention"].head() #Customers in Need Attention segment

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12360             51          3   2662.06            3              3   
12372             70          3   1298.04            3              3   
12413             65          3    758.10            3              3   
12456             43          3   3181.04            3              3   
12536             42          3  12601.83            3              3   

            MonetaryScore RFM_SCORE         Segment  
Customer ID                                          
12360                   5       335  Need Attention  
12372                   4       334  Need Attention  
12413                   3       333  Need Attention  
12456                   5       335  Need Attention  
12536                   5       335  Need Attention

In [131]:
rfm[rfm["Segment"] == "Loyal Customers"].index # Customer IDs of Loyal Customers

Int64Index([12352, 12359, 12370, 12380, 12388, 12395, 12407, 12408, 12415,
            12421,
            ...
            18204, 18211, 18221, 18226, 18235, 18236, 18257, 18259, 18263,
            18287],
           dtype='int64', name='Customer ID', length=820)

In [132]:
loyal_df = pd.DataFrame()

In [133]:
loyal_df["LoyalCustomerID"] = rfm[rfm["Segment"] == "Loyal Customers"].index

In [134]:
loyal_df.head()

LoyalCustomerID
0            12352
1            12359
2            12370
3            12380
4            12388

In [136]:
loyal_df.to_csv("loyal_customers.csv", index = False) # save Customers as dataframe to share with people

In [137]:
rfm1 = rfm[["Segment","Recency","Frequency", "Monetary"]].groupby("Segment").agg(["mean","median","count"])

In [138]:
rfm1.columns

MultiIndex([(  'Recency',   'mean'),
            (  'Recency', 'median'),
            (  'Recency',  'count'),
            ('Frequency',   'mean'),
            ('Frequency', 'median'),
            ('Frequency',  'count'),
            ( 'Monetary',   'mean'),
            ( 'Monetary', 'median'),
            ( 'Monetary',  'count')],
           )

In [140]:
rfm1.sort_values(( 'Monetary', 'median'), ascending = False) # descriptive statistics of RFM values for all segments

Recency              Frequency              Monetary  \
                       mean median count      mean median count     mean   
Segment                                                                    
Champions              4.37      3   632     12.34      8   632  6866.78   
Can't Loose          130.05    105    64      8.31      7    64  2791.01   
Loyal Customers       31.69     28   820      6.44      5   820  2862.89   
At Risk              152.04    137   594      2.87      3   594  1079.81   
Need Attention        50.27     50   184      2.32      2   184   894.49   
Potential Loyalists   15.37     16   486      2.01      2   486  1041.34   
About to Sleep        51.37     51   353      1.16      1   353   472.55   
Hibernating          215.66    217  1069      1.10      1  1069   488.86   
New Customers          5.43      6    42      1.00      1    42   388.21   
Promising             21.42     21    95      1.00      1    95   290.91   

                                   
                     median count  
Segment                            
Champions           2613.85   632  
Can't Loose         2236.74    64  
Loyal Customers     1737.83   820  
At Risk              675.16   594  
Need Attention       637.78   184  
Potential Loyalists  524.00   486  
About to Sleep       331.95   353  
Hibernating          293.00  1069  
New Customers        274.50    42  
Promising            219.00    95

In [141]:
rfm[["Segment","Recency","Frequency", "Monetary"]].groupby("Segment").agg(["mean","median","count"]).loc[["Can't Loose", "Loyal Customers", "Need Attention"]]

Recency              Frequency              Monetary          \
                   mean median count      mean median count     mean  median   
Segment                                                                        
Can't Loose      130.05    105    64      8.31      7    64  2791.01 2236.74   
Loyal Customers   31.69     28   820      6.44      5   820  2862.89 1737.83   
Need Attention    50.27     50   184      2.32      2   184   894.49  637.78   

                       
                count  
Segment                
Can't Loose        64  
Loyal Customers   820  
Need Attention    184

In [142]:
df["TotalPrice"].sum() #Tpotal money gained

8911425.904

In [143]:
Cl_indexes = rfm[rfm["Segment"] == "Can't Loose"].index #Can't Loose segment customers indexes

In [144]:
Cl_indexes_list = Cl_indexes.to_list() #cast to list 

In [145]:
df_cantloose = df[df["Customer ID"].isin(Cl_indexes_list)]  # recorda of customers in Can't Loose segment 

In [148]:
df_cantloose.TotalPrice.sum() #Total Money comes from this segment 

178624.68

In [149]:
rfm["Segment"].unique() #All segments

array(['Hibernating', 'Champions', 'At Risk', 'Promising',
       'Loyal Customers', 'Potential Loyalists', 'About to Sleep',
       'Need Attention', 'New Customers', "Can't Loose"], dtype=object)

In [150]:
rfm.shape[0]

4339

In [167]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE      Segment  
Customer ID                                       
12346                   5       115  Hibernating  
12347                   5       555    Champions  
12348                   4       244      At Risk  
12349                   4       414    Promising  
12350                   2       112  Hibernating

In [168]:
rfm[rfm["Segment"] == 'Hibernating'].shape[0] # customer count in Hibernating segment

1069

In [169]:
for seg in rfm["Segment"].unique():
    seg_indexes_list = rfm[rfm["Segment"] == seg].index.to_list()
    df_seg = df[df["Customer ID"].isin(seg_indexes_list)]
    print(seg, ' Segment Customer count: ', rfm[rfm["Segment"] == seg].shape[0])
    print(seg, ' Segment Customer Count / Total Customer Count Ratio: %.2f' %(rfm[rfm["Segment"] == seg].shape[0] / rfm.shape[0]))
    print('Money comes from ', seg, ' Segment Customers: %.2f' %df_seg.TotalPrice.sum(), 'pound')
    print('Ratio of money comes from ', seg, 'segment to Total Customers: %.2f' %(df_seg.TotalPrice.sum() / df["TotalPrice"].sum()))
    print('\n')
print('Elde edilen toplam gelir: %.2f' %df["TotalPrice"].sum()) #kazanilan toplam para
print('Toplam musteri sayisi: ', rfm.shape[0])

Hibernating  Segment Customer count:  1069
Hibernating  Segment Customer Count / Total Customer Count Ratio: 0.25
Money comes from  Hibernating  Segment Customers: 522589.96 pound
Ratio of money comes from  Hibernating segment to Total Customers: 0.06


Champions  Segment Customer count:  632
Champions  Segment Customer Count / Total Customer Count Ratio: 0.15
Money comes from  Champions  Segment Customers: 4339804.46 pound
Ratio of money comes from  Champions segment to Total Customers: 0.49


At Risk  Segment Customer count:  594
At Risk  Segment Customer Count / Total Customer Count Ratio: 0.14
Money comes from  At Risk  Segment Customers: 641409.59 pound
Ratio of money comes from  At Risk segment to Total Customers: 0.07


Promising  Segment Customer count:  95
Promising  Segment Customer Count / Total Customer Count Ratio: 0.02
Money comes from  Promising  Segment Customers: 27636.75 pound
Ratio of money comes from  Promising segment to Total Customers: 0.00


Loyal Customers  Seg

# Yorumlar

Bir tane çok para getiren, bu nedenle de kesinlikle kaybedilmemesi gereken (Can't loose), bir tane bize son derece bağlı olan (Loyal Customers), bir tane de kararsız olan, bu nedenle de dikkatli olmamızı gerektiren segment seçtim. (Need attention)

### Can't loose
- Bu grupta 64 musterimiz var.
- Az sayıda musteri olmasına rağmen,  herbir kisiden saglanan gelir dusunuldugunde cok onemli oldugu goruluyor. (median 2236 pound ve bu median degeri segmentteki kişilerin ortalama sagladiklari gelir dusunuldugunde en yuksek ikinci rakam.)
- Musteri sayisina gore dusunuldugunde elde edilen gelir gercekten hatiri sayilir. Champions segmenti ile kiyaslama yapidiginda bile, Champions'dan elde edilen gelirin yaklasik %4o'inin %10'u kadar kisi ile elde edildigi goruluyor. 
- median Recency degerleri (134) frequency degerleri ile beraber dusunuldugunde, monetary degerlerinin son derece yuksek oldugu goruluyor, yani kisi sayisi oldukca dusuk (64), ancak buna karsin kisi basina elde edilen medyan geir son derece yuksek (2236), bundan su sonucu cikariyorum;  eger bu segmentte bulunan musteri sayisini arttirirsak gelirlerimize katkisi yuksek olacaktir. Bunun icin siklikla kendimizi hatırlatmali bu segment profillerine gore promosyonlar duzenlemeli ve bu segmentin es vebu segmentte dostlarini musterimiz yapmaliyiz.
- Bu segmentteki musterilerin oldukca detayli profil analizleri yapilarak ayni profilde musterileri bu segmente dahi etmeye calismaliyiz.

### Loyal Customers
- 820 musterimiz var segnmentte.
- Her 5 musterimizden birinin obu segmentte olduğu ve yine gelirlerin yaklaşık %25'inin (2347570.48 sterlin) bu müşterilere ait olduğu düsünüldügünde segmentin onemi açıkça gorulecektir.
- Recency degerlerinin mean ve median degerleri diger segmentlere kiyasla oldukca iyi durumda. (median 28)
- Frequency degerleri ortalama yaklasik olarak 6, diger segmentlere gore iyi bir sonuc bu da
- Monetary degerleri incelendiginde,  mean 2862, median ise 1737. Burada pozitif bir carpiklik soz konusu monetary dagiliminda.
- Burada sadik musteriler ile ilgili genel bir yorumum var benim. Pazarlama ve Isletme konularinda oldukca sik genel bir kuraldan bahsedilir, o da sudur: halihazirdaki, yani kazanilmis musterileri elde tutmak yeni musteri kazanmaktan daha degerlidir. Uygulamada her zaman bu ornekteki gibi bariz olmasa da her zaman onemi gorulebilir. Bu ornekte gelirin %25'înin bu musterilerden elde edildigi dusunulurse, onlari elde tutmaya devam etmek icin ozel promosyonlar, kampanyalar, indirimler, tavsiyeler ve hediyelerle desteklenmeliler kesinlikle 

### Need Attention
- Segmentte 184 musteri mevcut.
- Elde edilen gelir 64586.21 pound ve toplam gelirin  %2'si, bu nedenle bu segment musterileri uzerlerine fazlaca dusmeye gerek olmadigini dusunduruyur.
- Recency median degeri 50, bu oldukca dusuk RFM score'unun dusuk olmasinin onemli bir nedeni bu. 
- Her ne kadar bu segmentten elde düedilen gelir oldukca dusuk olsa da,  184 musteri sayisi azımsanacak bir deger degil ve recency skorlarini yukseltmek icin ilgi cekici smsler gonderilebilir. Daha cok ziyaret etmelerini ve sitemizden urun alimi yapmalarini saglamak icin, daha once alim yaptıklari urunler ve beraber satilan ve musterin beraber tercih ettikleri urunler onerilebilir.

Aslinda, tum segmentteki musteriler icin genel olarak bir Birliktelik kuralları analizi calismasi yapilmali ve segmentlerdeki profillere ozel birliktelik kurallari cikarilarak incelenmeli ve bu surec otomatize edilerek surdurulebilir fayda sağlanmali diye dusunuyorum.







# Comments

I chose a segment that brings a lot of money and therefore should not be lost (Can't loose), one is very loyal to us (Loyal Customers), and one is unstable, so we need to be careful. (Need attention)

### Can't loose
- We have 64 customers in this group.
- Despite the small number of customers, it seems to be very important when considering the income from each person. (The median is 2236 pounds, and this median value is the second highest figure when considering the average income earned by people in the segment.)
- Considering the number of customers, the earned income is considered to be truly memorable. Even when comparing with the Champions segment, about 4% of revenues from Champions are generated by 10% of people.
- Considering the median recency values ​​(134) together with the frequency values, the monetary values ​​seem to be extremely high, i.e. the number of people is quite low (64), but the median income per person is extremely high (2236), I conclude from this ; If we increase the number of customers in this segment, its contribution to our revenues will be high. For this, we should remind ourselves frequently, we should organize promotions according to the profiles of this segment and make the friends of this segment as our customers.
- We should try to include customers with the same profile in this segment by making very detailed profile analysis of the customers in this segment.

### Loyal Customers
- We have 820 customers in the segment.
- Considering that one of our 5 customers is in the obu segment and approximately 25% of the revenues (£ 2347570.48) belongs to these customers, the importance of the segment will be clearly seen.
- The mean and median values ​​of the recency values ​​are very good compared to the other segments. (median 28)
- Frequency values ​​are about 6, a good result compared to other segments
- When the monetary values ​​are examined, the mean is 2862, the median is 1737. There is a positive twist here in the monetary distribution.
- Here I have a general comment about loyal customers. A general rule of thumb is often mentioned in Marketing and Business matters, which is: retaining existing, acquired customers, is worth more than acquiring new customers. In practice, it is always important, although not as obvious as in this example. Given that 25% of the revenue in this example is derived from these customers, they must be supported by special promotions, offers, discounts, recommendations and gifts to keep them retained.

### Need Attention
- There are 184 customers in the segment.
- The revenue generated is £ 64586.21 and 2% of the total revenue, so this segment does not think it is necessary to overthrow their customers.
- The recency median is 50, which is an important reason for the very low RFM score.
- Although the revenue earned from this segment is quite low, the number of 184 customers is not a small value and interesting sms can be sent to increase the recency scores. In order to enable them to visit more and buy products from our site, the products they bought before and the products that were sold together and the products that the customer preferred together can be recommended.

In fact, an association rules analysis should be done for customers in all segments in general, and special association rules should be examined by extracting special association rules for the profiles in the segments, and I think that this process should be automated to provide sustainable benefits.